HADOOP-17981 Support etag-assisted renames in FileOutputCommitter

steveloughran opened the pull request:
Contains MAPREDUCE-7367. #2971
Parallelize file moves in FileOutputCommitter v1 job commit

Contributed by Rajesh Balamohan


### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [x] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?

Commit:
MAPREDUCE-7367. Parallelize file moves in FileOutputCommitter v1 job commit

Contributed by Rajesh Balamohan

Change-Id: I8d2cedb0ba8b6aba2f1d0f814f52d77b156f1206

steveloughran committed:
HADOOP-17981. ResilientCommitByRename

Selective cherrypick of MAPREDUCE-7341 files; move etag tests
into hadoop-azure.
RawLocalFS implements the API; checksum fs does not

Change-Id: Ic68337742de6aefd067eda14313fd07088c73ac7

mukund-thakur commented on hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/committer/manifest/impl/StoreOperationsThroughFileSystem.java:
```diff
@@ -0,0 +1,271 @@
+/**
+ * Implement task and job operations through the filesystem API.
+ */
+public class StoreOperationsThroughFileSystem extends StoreOperations {
```
This class is not supposed to be a part of this commit right?

steveloughran replied:
I screwed up that commit by cherrypicking from the manifest committer branch (which I needed to) -but missed that git adds that as a file. Removed them

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m 30s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 6 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  13m 27s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  30m 36s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  33m 57s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  28m 42s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   4m 57s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   5m 25s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   3m 42s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   4m 36s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   7m 38s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 35s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 28s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 46s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  26m  9s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  26m  9s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  22m 14s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |  22m 14s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   3m 59s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/4/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 19 new + 99 unchanged - 0 fixed = 118 total (was 99)  |
| +1 :green_heart: |  mvnsite  |   4m  9s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   2m 56s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   3m 23s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   7m 21s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  23m 49s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  19m 57s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   6m 19s |  |  hadoop-mapreduce-client-core in the patch passed.  |
| +1 :green_heart: |  unit  | 143m 58s |  |  hadoop-mapreduce-client-jobclient in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 38s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  |  The patch does not generate ASF License warnings.  |
|  |   | 434m 32s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3597 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux 6ba9a731eb8d 4.15.0-143-generic #147-Ubuntu SMP Wed Apr 14 16:10:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / f4950b3d0416b764433d18d24e8b19b09b663cc4 |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/4/testReport/ |
| Max. process+thread count | 3143 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient hadoop-tools/hadoop-azure U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/4/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-17981. resilient commit.

adds options vararg similar to rename/3; allows caller to declare it is confident the
dest path doesn't exist, so no need to check for overwrites or similar.

test on local fs verifies that renaming on a filesystem without the new
interface matches that of the new behaviour,

committer test alternates between raw/local fs on parameterized runs.
minor reorg of the new tests in that class

Change-Id: Ic31d52c9a95b768e0a97583e906c20c53b120194

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 41s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 6 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  12m 49s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 18s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  21m 38s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  19m  0s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   3m 40s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   4m 26s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   3m 20s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   3m 52s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   6m 10s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m 27s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 28s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 31s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m  2s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  21m  2s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m  7s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |  19m  7s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   3m 37s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/5/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 17 new + 99 unchanged - 0 fixed = 116 total (was 99)  |
| +1 :green_heart: |  mvnsite  |   4m 21s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   3m 20s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   3m 49s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   6m 51s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  20m 36s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m 21s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   5m 57s |  |  hadoop-mapreduce-client-core in the patch passed.  |
| +1 :green_heart: |  unit  | 135m 38s |  |  hadoop-mapreduce-client-jobclient in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 41s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 12s |  |  The patch does not generate ASF License warnings.  |
|  |   | 370m 13s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3597 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux 80640d7e2e23 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 712c7cbc0dad45bf31378ec38935dc6141faff2e |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/5/testReport/ |
| Max. process+thread count | 1573 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient hadoop-tools/hadoop-azure U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/5/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

sidseth commented on hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/FileOutputCommitter.java:
```diff
@@ -497,6 +535,302 @@ private void mergePaths(FileSystem fs, final FileStatus from,
+      Path finalOutput = getOutputPath();
+      FileSystem fs = finalOutput.getFileSystem(context.getConfiguration());
+      // created resilient commit helper bonded to the destination FS/path
+      resilientCommitHelper = new ResilientCommitByRenameHelper(fs);
```
This mechanism becomes very FileSystem specific. Implemented by Azure right now.
Other users of rename will not see the benefits without changing interfaces, which in turn requires shimming etc.

Would it be better for AzureFileSystem rename itself to add a config parameter which can lookup the src etag (at the cost of a performance hit for consistency), so that downstream components / any users of the rename operation can benefit from this change without having to change interfaces. Also, if the performance penalty is a big problem - Abfs could create very short-lived caches for FileStatus objects, and handle errors on discrepancies with the cached copy.
Essentially - don't force usage of the new interface to get the benefits.

Side note: The fs.getStatus within ResilientCommitByRenameHelper for FileSystems where this new functionality is not supported will lead to a performance penalty for the other FileSystems (performing a getFileStatus on src).

mukund-thakur commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/ResilientCommitByRename.java:
nit : typo.. Flags.

mukund-thakur reviewed (commented):
Looks good overall. Pending ABFS side changes.

steveloughran commented:
> This mechanism becomes very FileSystem specific. Implemented by Azure right now.

I agree, which is why the API is restricted for its uses to mr-client-core only.
as abfs is the only one which needs it for correctness under load,
And I'm not worried about that specifity.
Can I point to how much of the hadoop fs api are hdfs-only -and they are public.

> Other users of rename will not see the benefits without changing interfaces, which in turn requires shimming etc.

Please don't try and use this particular interface in Hive. 

> Would it be better for AzureFileSystem rename itself to add a config parameter which can lookup the src etag (at the cost of a performance hit for consistency), so that downstream components / any users of the rename operation can benefit from this change without having to change interfaces.

We are going straight from a listing (1 request/500 entries) to that rename. doing a HEAD first cuts the throughtput in half. so no.

> Also, if the performance penalty is a big problem - Abfs could create very short-lived caches for FileStatus objects, and handle errors on discrepancies with the cached copy.
 
Possible but convoluted.
 
 
> Essentially - don't force usage of the new interface to get the benefits.

I understand the interests of the hive team, but this fix is not the place to do a better API.

Briefly cacheing the source FS entries is something to consider though. Not this week.

What I could do with is some help getting #2735 in, then we can start on a public rename() builder API which will take a file status, as openFile does.

> Side note: The fs.getStatus within ResilientCommitByRenameHelper for FileSystems where this new functionality is not supported will lead to a performance penalty for the other FileSystems (performing a getFileStatus on src).

There is an option to say "i know it is not there"; this skips the check. the committer passes this option down because it issues a delete call first.
 
FWIW the manifest committer will make that pre-rename commit optional, saving that IO request. I am curious as to how well that will work I went executed on well formed tables.

steveloughran committed:
HADOOP-17981. changes during a review with Mukund

Change-Id: I65dd20e2e0d6af2d347c07b4a3e13ab58763c1ac

steveloughran committed:
HADOOP-17981. add a test I'd forgotten to commit

Change-Id: I79297a308bf558d188592cadbe60e2622435c6e8

steveloughran commented:
@sidseth I discussed the idea of caching status etags with @mukund-thakur 

considering that you'd need to remove entries from the cache on: create, rename, & delete, it would actually be a copy of `org.apache.hadoop.fs.s3a.s3guard.LocalMetadataStore`, as #3534 is removing all of s3guard, I'll consider that a WONTFIX. sorry.

steveloughran commented:
see #3611 . i think this pr had got too complex and so have gone back to what I had originally done in the manifest committer. filesystem provides etags, application code recovers.

This is something which we can expose to the hive team, as all that will be added is a single stable extension to filestatus to get the etag.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 44s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 7 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  12m 44s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 26s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  21m 42s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  19m  0s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   3m 37s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   4m 23s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   3m 22s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   3m 48s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   6m 12s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m 26s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 28s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 34s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 16s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  21m 16s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 58s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |  19m 58s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   3m 31s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/6/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 13 new + 99 unchanged - 0 fixed = 112 total (was 99)  |
| +1 :green_heart: |  mvnsite  |   4m  3s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   3m  4s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   3m 40s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   6m 54s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  20m 46s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m 29s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   5m 58s |  |  hadoop-mapreduce-client-core in the patch passed.  |
| +1 :green_heart: |  unit  | 134m 34s |  |  hadoop-mapreduce-client-jobclient in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 33s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 11s |  |  The patch does not generate ASF License warnings.  |
|  |   | 369m  9s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3597 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux bdd11688ebd6 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 72bc6fd23f78bbe2f33491ecd075a314d390adf8 |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/6/testReport/ |
| Max. process+thread count | 1325 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient hadoop-tools/hadoop-azure U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3597/6/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

sidseth commented:
> > This mechanism becomes very FileSystem specific. Implemented by Azure right now.
> 
> I agree, which is why the API is restricted for its uses to mr-client-core only. as abfs is the only one which needs it for correctness under load, And I'm not worried about that specifity. Can I point to how much of the hadoop fs api are hdfs-only -and they are public.
> 
> > Other users of rename will not see the benefits without changing interfaces, which in turn requires shimming etc.
> 
> Please don't try and use this particular interface in Hive.
> 
Was referring to any potential usage - including Hive.
> > Would it be better for AzureFileSystem rename itself to add a config parameter which can lookup the src etag (at the cost of a performance hit for consistency), so that downstream components / any users of the rename operation can benefit from this change without having to change interfaces.
> 
> We are going straight from a listing (1 request/500 entries) to that rename. doing a HEAD first cuts the throughtput in half. so no.
> 
In the scenario where this is encountered. Would not be the default behaviour, and limits the change to Abfs. Could also have the less consistent version which is not etag based, and responds only on failures. Again - limited to Abfs.
> > Also, if the performance penalty is a big problem - Abfs could create very short-lived caches for FileStatus objects, and handle errors on discrepancies with the cached copy.
> 
> Possible but convoluted.
> 
Agree. Quite convoluted. Tossing in potential options - to avoid a new public API.
> > Essentially - don't force usage of the new interface to get the benefits.
> 
> I understand the interests of the hive team, but this fix is not the place to do a better API.
> 
> Briefly cacheing the source FS entries is something to consider though. Not this week.
> 
> What I could do with is some help getting #2735 in, then we can start on a public rename() builder API which will take a file status, as openFile does.
> 
This particular change would be FSImpl agnostic, and potentially remove the need for the new interface here?
> > Side note: The fs.getStatus within ResilientCommitByRenameHelper for FileSystems where this new functionality is not supported will lead to a performance penalty for the other FileSystems (performing a getFileStatus on src).
> 
> There is an option to say "i know it is not there"; this skips the check. the committer passes this option down because it issues a delete call first.
> 
EOD - this ends up being a new API (almost on the FileSystem), which is used by the committer first; then someone discovers it and decides to make use of it.
> FWIW the manifest committer will make that pre-rename commit optional, saving that IO request. I am curious as to how well that will work I went executed on well formed tables.

steveloughran commented:
> EOD - this ends up being a new API (almost on the FileSystem), which is used by the committer first; then someone discovers it and decides to make use of it.

yes, and both hive and hbase are known to do that, often ending forcing hdfs only apis to get pulled up without concern for the other stores (see git history of FileSystem there). a builder rename would be a big job with
1. rename/3 public with tests and performant object store impls
2. default builder to delegate to that with same semantics
3. add options for etag etc on a store by store or cross store basis
